src/Primitives/src/ChangeToken.cs

```diff
@@ -67,26 +50,99 @@ public static IDisposable OnChange<TState>(Func<IChangeToken> changeTokenProduce
+                _changeTokenConsumer = changeTokenConsumer;
+                _state = state;
+
+                var token = changeTokenProducer();
+
+                RegisterChangeTokenCallback(token);
+            }
+
+            private void OnChangeTokenFired()
             {
                 // The order here is important. We need to take the token and then apply our changes BEFORE
                 // registering. This prevents us from possible having two change updates to process concurrently.
                 //
                 // If the token changes after we take the token, then we'll process the update immediately upon
                 // registering the callback.
-                var t = changeTokenProducer();
+                var token = _changeTokenProducer();
+
                 try
                 {
-                    changeTokenConsumer((TState)s);
+                    _changeTokenConsumer(_state);
                 }
-                finally // We always want to ensure the callback is registered
+                finally
                 {
-                    t.RegisterChangeCallback(callback, s);
+                    // We always want to ensure the callback is registered
+                    RegisterChangeTokenCallback(token);
                 }
-            };
+            }
+
+            private void RegisterChangeTokenCallback(IChangeToken token)
+            {
+                var registraton = token.RegisterChangeCallback(s => ((ChangeTokenRegistration<TState>)s).OnChangeTokenFired(), this);
+
+                SetDisposable(registraton);
+            }
+
+            private void SetDisposable(IDisposable disposable)
+            {
+                // We don't want to transition from _disposedSentinel => anything since it's terminal
+                // but we want to allow going from previously assigned disposable, to another
+                // disposable. The assumption here is that we don't have to dispose the registration if an IChangeToken
+                // fires.
+                var current = Volatile.Read(ref _disposable);
+
+                // If Dispose was called, then immediately dispose the disposable
+                if (current == _disposedSentinel)
+                {
+                    disposable.Dispose();
+                    return;
+                }
+
+                // Otherwise, try to update the disposable
+                var previous = Interlocked.CompareExchange(ref _disposable, disposable, current);
 
-            return changeTokenProducer().RegisterChangeCallback(callback, state);
+                if (previous == _disposedSentinel)
+                {
+                    // The subscription was disposed so we dispose immediately and return
+                    disposable.Dispose();
+                }
+                else if (previous == current)
+                {
+                    // We successfuly assigned the _disposable field to disposable
```

Comment:
You don't need to dispose of the previous one if it's non-null?

Reply:
I was assuming once he callback is triggered there’s no need to dispose the previous token (like a CancellationTokenRegistration). Maybe it still make sense to always dispose it.

Reply:
This seems to break things, I'll revert then investigate